src/editors/documentController.ts

```diff
@@ -0,0 +1,262 @@
+          }
+
+          if (!documents || documents.length === 0) {
+            const errorMessage = `Unable to find document: ${JSON.stringify(
+              documentId
+            )}`;
+
+            vscode.window.showErrorMessage(errorMessage);
+
+            return reject(new Error(errorMessage));
+          }
+
+          return resolve(JSON.parse(EJSON.stringify(documents[0])));
+        }
+      );
+    });
+  }
+
+  _saveDocumentFailed(errorMessage: string): boolean {
+    // Send a telemetry event that saving the document failed.
+    this._telemetryController.trackDocumentUpdated('treeview', false);
+
+    vscode.window.showErrorMessage(errorMessage);
+
+    return false;
+  }
+
+  async saveDocumentToMongoDB(): Promise<boolean> {
+    const activeEditor = vscode.window.activeTextEditor;
+
+    log.info('save document to MongoDB', activeEditor);
+
+    if (!activeEditor) {
+      return false;
+    }
+
+    const uriParams = new URLSearchParams(activeEditor.document.uri.query);
+    const documentLocation =
+      uriParams.get(DOCUMENT_LOCATION_URI_IDENTIFIER) || '';
+    const namespace = uriParams.get(NAMESPACE_URI_IDENTIFIER);
+    const connectionId = uriParams.get(CONNECTION_ID_URI_IDENTIFIER);
+    const documentIdReference = uriParams.get(DOCUMENT_ID_URI_IDENTIFIER) || '';
+    const documentId = this._documentIdStore.get(documentIdReference);
+
+    // If not MongoDB document save to disk instead of MongoDB.
+    if (
+      documentLocation !== 'mongodb' ||
+      !namespace ||
+      !connectionId ||
+      !documentId
+    ) {
+      vscode.commands.executeCommand('workbench.action.files.save');
+
+      return false;
+    }
+
+    const activeConnectionId = this._connectionController.getActiveConnectionId();
+    const connectionName = this._connectionController.getSavedConnectionName(
+      connectionId
+    );
+
+    if (activeConnectionId !== connectionId) {
+      return this._saveDocumentFailed(
+        `Unable to save document: no longer connected to '${connectionName}'`
+      );
+    }
+
+    const dataservice = this._connectionController.getActiveDataService();
+
+    if (dataservice === null) {
+      return this._saveDocumentFailed(
+        `Unable to save document: no longer connected to '${connectionName}'`
+      );
+    }
+
+    return await this.replaceDocument(
+      documentId,
+      namespace,
+      dataservice,
+      activeEditor
+    );
+  }
+
+  async replaceDocument(
+    documentId: EJSON.SerializableTypes,
+    namespace: string,
+    dataservice: DataServiceType,
+    editor: vscode.TextEditor
+  ): Promise<boolean> {
+    const activeEditor = vscode.window.activeTextEditor;
+    let newDocument: EJSON.SerializableTypes = {};
+
+    try {
+      newDocument = EJSON.parse(editor.document.getText());
+    } catch (error) {
+      return this._saveDocumentFailed(error.message);
+    }
+
+    this._statusView.showMessage('Saving document...');
+
+    dataservice.findOneAndReplace(
```

Comment:
Should we promisify this?
```ts
const findOneAndReplace = util.promisify(dataservice.findOneAndReplace.bind(dataservice));

try {
  await findOneAndReplace(...
```
Does the `replaceDocument` function return before this completes?
Do we have a test that uses a database for this function? Might be good for ensuring we don't introduce any regressions in future.

Reply:
Replaced callbacks with `util.promisify` and added a test from `replaceDocument `.